controllers/storagecluster/external_ocs.go

```diff
@@ -0,0 +1,221 @@
+package storagecluster
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	externalClient "github.com/red-hat-storage/ocs-operator/services/provider/client"
+)
+
+var (
+	// externalOCSResources will hold the ExternalResources for storageclusters
+	// ExternalResources can be accessible using the UID of an storagecluster
+	externalOCSResources = map[types.UID][]ExternalResource{}
+)
+
+// isExternalOCSProvider returns true if it is ocs to ocs ExternalStorage consumer cluster
+func isExternalOCSProvider(instance *ocsv1.StorageCluster) bool {
+	return instance.Spec.ExternalStorage.Enable && instance.Spec.ExternalStorage.StorageProviderKind == ocsv1.KindOCS
+}
+
+// newExternalClusterClient returns the *grpc.ClientConn and *providerClient.ConsumerClient
+func newExternalClusterClient(instance *ocsv1.StorageCluster) (*grpc.ClientConn, *externalClient.OCSProviderClient, error) {
+
+	// TODO: implement logic for separating address and token
+
+	grpcConnection, err := externalClient.NewGRPCConnection(
+		instance.Spec.ExternalStorage.ConnectionString, []grpc.DialOption{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	consumerClient := externalClient.NewProviderClient(grpcConnection, time.Second*10)
+	return grpcConnection, consumerClient, nil
+}
+
+// onboardConsumer makes an API call to the external storage provider cluster for onboarding
+func onboardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	clusterVersion := &configv1.ClusterVersion{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: "version"}, clusterVersion)
+	if err != nil {
+		r.Log.Error(err, "External-OCS:Failed to get the clusterVersion version of the OCP cluster")
+		return reconcile.Result{}, err
+	}
+
+	name := fmt.Sprintf("ocs-consumer-%s", clusterVersion.Spec.ClusterID)
+	// TODO: implement logic for separating address and token
+	response, err := r.externalClusterClient.OnboardConsumer(
+		instance.Spec.ExternalStorage.ConnectionString, name,
+		instance.Spec.ExternalStorage.RequestedCapacity.String())
+	if err != nil {
+		s, ok := status.FromError(err)
+		if !ok {
+			// This is not gRPC error. The operation must have failed before gRPC
+			// method was called, otherwise we would get gRPC error.
+			return reconcile.Result{}, err
+		}
+		r.logGrpcErrorAndReportEvent(instance, "OnboardConsumer", err, s.Code())
+		return reconcile.Result{}, err
+	}
```

Comment:
Why do we care about the origin of the error? 

But if you really od care pelase use the reversed pattern as it is shorter and much easier to read:
```suggestion
	if err != nil {
		if s, ok := status.FromError(err), ok {
			 r.logGrpcErrorAndReportEvent(instance, "OnboardConsumer", err, s.Code())
		}
		return reconcile.Result{}, err
	}
```

This comment should apply to all of the instances of the original pattern below.

Reply:
done